### ISSUE TITLE
Websocket example doesn't seem to work with Erlang 17.4r2.

### ISSUE THREAD

**Issue by ubill** [ninenines/cowboy#786]
Hi, websocket project does not seem to work when I upgraded my Erlang to the latest 17.4 rev.2.
My system is Debian 7.8.
I downloaded the Erlang 17.4 r2 from the erlang-solutions site.

What happens is init/2 in ws_handler no longer works:

I believe this line says it: [{reason,undef},{mfa,{ws_handler,init,3}

Because in ws_handler, we only have init/2 but somehow, just trying to load the localhost:8090 in the browser will trigger this error. Full trace included (I couldn't make any sense of it.)

I reverted my VM's snapshot back to when I had Erlang 17.3.2 and it worked again.

My cowboy is the latest master, and cowlib and ranch are the versions obtained with running make on cowboy.

That's what I could report to the best of my limited understanding. Hope you could reproduce it easily and find the root cause, cheers.

```
Eshell V6.3  (abort with ^G)
(websocket_example@127.0.0.1)1> 
=ERROR REPORT==== 30-Jan-2015::22:50:17 ===
Ranch listener http had connection process started with cowboy_protocol:start_link/4 at <0.164.0> exit with reason: {[{reason,undef},{mfa,{ws_handler,init,3}},{stacktrace,[{ws_handler,init,[{tcp,http},{http_req,#Port<0.626>,ranch_tcp,keepalive,<0.164.0>,<<"GET">>,'HTTP/1.1',{{127,0,0,1},45776},<<"localhost">>,undefined,8080,<<"/websocket">>,undefined,<<>>,undefined,[],[{<<"host">>,<<"localhost:8080">>},{<<"connection">>,<<"Upgrade">>},{<<"pragma">>,<<"no-cache">>},{<<"cache-control">>,<<"no-cache">>},{<<"upgrade">>,<<"websocket">>},{<<"origin">>,<<"http://localhost:8080">>},{<<"sec-websocket-version">>,<<"13">>},{<<"user-agent">>,<<"Mozilla/5.0 (X11; Linux x86_64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/40.0.2214.94 Safari/537.36">>},{<<"accept-encoding">>,<<"gzip, deflate, sdch">>},{<<"accept-language">>,<<"en-US,en;q=0.8">>},{<<"sec-websocket-key">>,<<"hg0sX0cX67C1m+ZC6Wp0Uw==">>},{<<"sec-websocket-extensions">>,<<"permessage-deflate; client_max_window_bits">>}],[{<<"connection">>,[<<"upgrade">>]}],undefined,[],waiting,<<>>,undefined,false,waiting,[],<<>>,undefined},[]],[]},{cowboy_handler,handler_init,4,[{file,"src/cowboy_handler.erl"},{line,64}]},{cowboy_protocol,execute,4,[{file,"src/cowboy_protocol.erl"},{line,442}]}]},{req,[{socket,#Port<0.626>},{transport,ranch_tcp},{connection,keepalive},{pid,<0.164.0>},{method,<<"GET">>},{version,'HTTP/1.1'},{peer,{{127,0,0,1},45776}},{host,<<"localhost">>},{host_info,undefined},{port,8080},{path,<<"/websocket">>},{path_info,undefined},{qs,<<>>},{qs_vals,undefined},{bindings,[]},{headers,[{<<"host">>,<<"localhost:8080">>},{<<"connection">>,<<"Upgrade">>},{<<"pragma">>,<<"no-cache">>},{<<"cache-control">>,<<"no-cache">>},{<<"upgrade">>,<<"websocket">>},{<<"origin">>,<<"http://localhost:8080">>},{<<"sec-websocket-version">>,<<"13">>},{<<"user-agent">>,<<"Mozilla/5.0 (X11; Linux x86_64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/40.0.2214.94 Safari/537.36">>},{<<"accept-encoding">>,<<"gzip, deflate, sdch">>},{<<"accept-language">>,<<"en-US,en;q=0.8">>},{<<"sec-websocket-key">>,<<"hg0sX0cX67C1m+ZC6Wp0Uw==">>},{<<"sec-websocket-extensions">>,<<"permessage-deflate; client_max_window_bits">>}]},{p_headers,[{<<"connection">>,[<<"upgrade">>]}]},{cookies,undefined},{meta,[]},{body_state,waiting},{buffer,<<>>},{multipart,undefined},{resp_compress,false},{resp_state,waiting},{resp_headers,[]},{resp_body,<<>>},{onresponse,undefined}]},{opts,[]}],[{cowboy_protocol,execute,4,[{file,"src/cowboy_protocol.erl"},{line,442}]}]}

=ERROR REPORT==== 30-Jan-2015::22:50:17 ===
Error in process <0.164.0> on node 'websocket_example@127.0.0.1' with exit value: {[{reason,undef},{mfa,{ws_handler,init,3}},{stacktrace,[{ws_handler,init,[{tcp,http},{http_req,#Port<0.626>,ranch_tcp,keepalive,<0.164.0>,<<3 bytes>>,'HTTP/1.1',{{127,0,0,1},45776},<<9 bytes>>,undefined,8080,<<10 bytes>>,undefined,<<0 bytes>>,undefined,[],[{<<4 bytes>>,<<14 bytes>>},{<<10 bytes>>,<<7 bytes>>},{<<6 bytes>>,<<8 bytes>>},{<<13 bytes>>,<<8 bytes>>},{<<7 bytes>>,<<9 bytes>>},{<<6 bytes>>,<<21 bytes>>},{<<21 bytes>>,<<2 bytes>>},{<<10 bytes>>,<<104 bytes>>},{<<15 bytes>>,<<19 bytes>>},{<<15 bytes>>,<<14 bytes>>},{<<17 bytes>>,<<24 bytes>>},{<<24 bytes>>,...
```

**Comment by essen** [comment]
Your Cowboy isn't the latest master when you use 17.4, because there is no init/3 callback anymore. Try to find why it gets the wrong Cowboy version.

**Comment by ubill** [comment]
Just to report my current version now when everything is working:
- Erlang 17.3.2
- cowlib 7d8a571 (tag 1.0.1)
- ranch 741bfb73 (tag 1.1.0)
- cowboy 02b1b281 (master)
and everything is working now.

I don't think I will try with 17.4 again soon as I just painstakingly got things working again now, but if you say you don't see it when you run with 17.4 then I guess it should be OK? And for now this issue could be closed I guess...

**Comment by ubill** [comment]
I tried to reproduce this problem again.

A question I have is: Are the versions of ranch and cowboy, somehow auto-selected/dependent on what version of Erlang I am using?

Steps to reproduce issue:
- When I am at Erlang version 17.3.2, this line
    {ok, _} = cowboy:start_http(http, 100, [{port, 8091}], [
        {env, [{dispatch, Dispatch}]}
    ]),
fails with error
** exception error: undefined function ranch:start_listener/6
     in function  main:start/2 (main.erl, line 13)
- I uninstalled Erlang and installed version 17.4.
Running the exact same thing now, just passes without any error.

However, browsing to my localhost:port page gets these errors:
=ERROR REPORT==== 15-Feb-2015::14:30:57 ===
Ranch listener http had connection process started with cowboy_protocol:start_link/4 at <0.155.0> exit with reason: {[{reason,undef},{mfa,{toppage_handler,init,3}},{stacktrace,[{toppage_handler,init,[{tcp,http},{http_req,#Port<0.1304>,ranch_tcp,keepalive,<0.155.0>,<<"GET">>,'HTTP/1.1'

=ERROR REPORT==== 15-Feb-2015::14:30:57 ===
Error in process <0.155.0> with exit value: {[{reason,undef},{mfa,{toppage_handler,init,3}},{stacktrace,[{toppage_handler,init,[{tcp,http},{http_req,#Port<0.1304>,ranch_tcp,keepalive,<0.155.0>,<<3 bytes>>,'HTTP/1.1',{{127,0,0,1},40427},<<9 bytes>>,undefined,8091,<<1 byte>>

That init/3 error again!
(I am using the toppage_handler with -export([init/2]) as from the cowboy/examples.)

My cowboy has been rebar compiled at these versions:
cowboy: 5ff3fdd
cowlib: 1.0.0
ranch: 1.0.0

I couldn't make sense of the problem.

**Comment by essen** [comment]
There's nothing looking for the Erlang version.

If the Cowboy in your deps is this version and it looks for init/3 then that means you have another Cowboy elsewhere in the path and that other Cowboy gets used instead of the one in deps.

**Comment by tuscland** [comment]
I am pretty sure you have the Erlang Solutions packaged Erlang OTP distribution.  

They recently included Cowboy, gproc and a bunch of utility "by popular request". 

 I stopped using their package for this reason.

**Comment by bullno1** [comment]
Try running `code:which(cowboy)` in an Erlang shell and see what it prints.

**Comment by ubill** [comment]
Thank you essen, tucsland, bullno1 for the insight! Indeed you are right - it is Erlang Solutions packaged Erlang, and it completely blindsided me that they package Cowboy together. code:which(cowboy) confirms this and it's 

code:which(cowboy).
"/usr/lib/erlang/lib/cowboy-1.0.1/ebin/cowboy.beam"

I will stop using the Erlang packaged by them and build the latest Erlang from source.

I will try it out right away and confirm if the problem goes away!

**Comment by ubill** [comment]
Update: I continued to use the Erlang Solutions packaged Erlang 17.4 .deb (because it seemed it would take longer, for the moment at least, to get the other dependencies like wxWidgets to build Erlang from source) - but now, I deleted out all the apps that are not from the Erlang standard libraries from the install path ( /usr/lib/erlang/lib ). Now the Cowboy and dependencies versions resolved and used are those from my local Git repo and are of the expected version. I am able to make and run Cowboy without any errors yet - at least I have not encountered any yet so far now.

Will close this ticket now.
